### PR TITLE
fix: flaky test `Sentry errors before initialization, after opting into metrics @no-mmi should send error events in background`

### LINKED ISSUE #26217: tests: `Sentry errors before initialization, after opting into metrics @no-mmi should send error events in background`
### What is this about?

https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/94145/workflows/fe416663-f2ed-4123-9bb8-3a87ec5845dc/jobs/3503745/tests

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

_No response_

### Stakeholder review needed before the work gets merged

- [ ] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

_No response_

### PR DESCRIPTION
**Pull request by seaona**
## **Description**
[The test is flaky](https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/94004/workflows/25f428d4-8fcb-4aca-8ddb-49edb51b57cf/jobs/3499122/tests) as sometimes we get a different error in the mocked endpoint, `Error: Could not establish connection. Receiving end does not exist`, instead of the expected one. This is because the condition for entering into the mock is not strict enough, and both errors include `value":"Cannot read properties of undefined (reading 'version')` in their body.

The solution here is:
- make the condition mock more strict, to make sure we capture the error we want (not the `receiving end` one)
- verify that this specific request is mocked - no need to assert on the json body as we already have this in the mock condition, however we leave it for debugging purposes. The change that has been made, is to assert the complete error all in once, so we get a more meaningful assertion and failure (if that's the case)

Screenshot from 2024-07-30 08-32-30

Case 1: the same mock, enters in the condition, with the error `Error","value":"Could not establish connection`, this will make the test fail.

```
{"exception":{"values":[{"type":"Error","value":"Could not establish connection. Receiving end does not exist.","stacktrace":{"frames":[{"filename":"/metamask/common-8.js","function":"wrappedSendMessageCallback","in_app":true,"lineno":15202,"colno":20}]},"mechanism":{"type":"onunhandledrejection","handled":false}}]},"level":"error","platform":"javascript","request":{"url":"/metamask/scripts/app-init.js","headers":{"User-Agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"}},"event_id":"350b00b955444903ae0e4572268eba8b","timestamp":1722323826.34,"environment":"development","release":"11.16.15","sdk":{"integrations":["InboundFilters","FunctionToString","BrowserApiErrors","Breadcrumbs","GlobalHandlers","LinkedErrors","Dedupe","HttpContext","FilterEvents","ExtraErrorData","BrowserProfiling"],"name":"sentry.javascript.browser","version":"8.19.0","packages":[{"name":"npm:@sentry/browser","version":"8.19.0"}]},"breadcrumbs":[{"timestamp":1722323826.33,"category":"sentry.event","event_id":"f84aa75ba52f450fa08c535372648a46","level":"error","message":"TypeError: Cannot read properties of undefined (reading 'version')"}],"contexts":{"trace":{"trace_id":"9dc46f6dfa2f42918276344dd54426f7","span_id":"b72b97bd6b5a7d97"},"Error":{}},"extra":{"appState":{"browser":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36","version":"11.16.15","persistedState":{"data":{"AuthenticationController":{"isSignedIn":"boolean"},"UserStorageController":{"isProfileSyncingEnabled":true},"MetamaskNotificationsController":{"subscriptionAccountsSeen":"object","isFeatureAnnouncementsEnabled":"boolean","isMetamaskNotificationsEnabled":"boolean","isMetamaskNotificationsFeatureSeen":"boolean","metamaskNotificationsList":"object","metamaskNotificationsReadList":"object"},"AccountsController":{"internalAccounts":{"selectedAccount":"string","accounts":"object"}},"AlertController":{"alertEnabledness":{"unconnectedAccount":true,"web3ShimUsage":true},"unconnectedAccountAlertShownOrigins":"object","web3ShimUsageOrigins":"object"},"AnnouncementController":{"announcements":"object"},"NetworkOrderController":{"orderedNetworkList":{"0":"object","1":"object","2":"object"}},"AccountOrderController":{"pinnedAccountList":{},"hiddenAccountList":{}},"AppStateController":{"browserEnvironment":{},"nftsDropdownState":{},"connectedStatusPopoverHasBeenShown":true,"termsOfUseLastAgreed":1722323821698,"defaultHomeActiveTabName":null,"fullScreenGasPollTokens":[],"notificationGasPollTokens":[],"popupGasPollTokens":[],"qrHardware":{},"recoveryPhraseReminderHasBeenShown":true,"recoveryPhraseReminderLastShown":1722323821698,"showTestnetMessageInDropdown":true,"trezorModel":null,"newPrivacyPolicyToastClickedOrClosed":"boolean","newPrivacyPolicyToastShownDate":"number","usedNetworks":{"0x1":true,"0xe708":true,"0x5":true,"0x539":true},"snapsInstallPrivacyWarningShown":true},"CurrencyController":{"currentCurrency":"usd","currencyRates":{"ETH":{"conversionDate":1665507600,"conversionRate":1700,"usdConversionRate":1700}}},"GasFeeController":{"estimatedGasFeeTimeBounds":{},"gasEstimateType":"none","gasFeeEstimates":{}},"KeyringController":{"vault":"string"},"MetaMetricsController":{"eventsBeforeMetricsOptIn":"object","fragments":"object","metaMetricsId":"fake-metrics-id","participateInMetaMetrics":true,"dataCollectionForMarketing":"boolean","traits":"object"},"NetworkController":{"selectedNetworkClientId":"string","networksMetadata":{"networkConfigurationId":{"EIPS":{},"status":"available"}},"providerConfig":{"chainId":"0x539","nickname":"Localhost 8545","rpcPrefs":"object","rpcUrl":"string","ticker":"ETH","type":"rpc","id":"networkConfigurationId"},"networkConfigurations":"object"},"OnboardingController":{"completedOnboarding":true,"firstTimeFlowType":"import","onboardingTabs":"object","seedPhraseBackedUp":true},"PermissionController":{"subjects":"object"},"PreferencesController":{"advancedGasFee":null,"currentLocale":"en","useExternalServices":"boolean","dismissSeedBackUpReminder":true,"featureFlags":{},"forgottenPassword":false,"identities":"object","ipfsGateway":"string","knownMethodData":"object","ledgerTransportType":"webhid","lostIdentities":"object","openSeaEnabled":false,"preferences":{"hideZeroBalanceTokens":false,"showExtensionInFullSizeView":false,"showFiatInTestnets":false,"showTestNetworks":false,"smartTransactionsOptInStatus":false,"useNativeCurrencyAsPrimaryCurrency":true,"petnamesEnabled":true,"showTokenAutodetectModal":"boolean","isRedesignedConfirmationsDeveloperEnabled":"boolean","showConfirmationAdvancedDetails":false},"selectedAddress":"string","theme":"light","useBlockie":false,"useNftDetection":false,"useNonceField":false,"usePhishDetect":true,"useTokenDetection":false,"useCurrencyRateCheck":true,"useMultiAccountBalanceChecker":true,"useRequestQueue":true},"QueuedRequestController":{"queuedRequestCount":0},"SelectedNetworkController":{"domains":"object"},"SmartTransactionsController":{"smartTransactionsState":{"fees":{},"liveness":true,"smartTransactions":"object"}},"SubjectMetadataController":{"subjectMetadata":"object"},"TokensController":{"allDetectedTokens":{},"allIgnoredTokens":{},"allTokens":{},"detectedTokens":"object","ignoredTokens":"object","tokens":"object"},"TransactionController":{"transactions":"object"},"config":"object","firstTimeInfo":"object"}}}}}
```

Case 2: the same mock, enters in the condition, with the error `"TypeError","value":"Cannot read properties of undefined (reading 'version')","stacktrace":`, this will make the test pass.

```
{"exception":{"values":[{"type":"TypeError","value":"Cannot read properties of undefined (reading 'version')","stacktrace":{"frames":[{"filename":"/metamask/background-5.js","function":"async initBackground","in_app":true,"lineno":36184,"colno":5},{"filename":"/metamask/background-5.js","function":"async initialize","in_app":true,"lineno":35538,"colno":22},{"filename":"/metamask/background-5.js","function":"loadStateFromPersistence","in_app":true,"lineno":35696,"colno":34},{"filename":"/metamask/background-0.js","function":"Migrator.migrateData","in_app":true,"lineno":12867,"colno":47},{"filename":"<anonymous>","function":"Array.filter","in_app":true},{"filename":"/metamask/background-0.js","function":"migrationIsPending","in_app":true,"lineno":12908,"colno":53}]},"mechanism":{"type":"onunhandledrejection","handled":false}}]},"level":"error","platform":"javascript","request":{"url":"/metamask/scripts/app-init.js","headers":{"User-Agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36"}},"event_id":"82038055aaf44c82b48b4b8c46fed6f8","timestamp":1722323813.849,"environment":"development","release":"11.16.15","sdk":{"integrations":["InboundFilters","FunctionToString","BrowserApiErrors","Breadcrumbs","GlobalHandlers","LinkedErrors","Dedupe","HttpContext","FilterEvents","ExtraErrorData","BrowserProfiling"],"name":"sentry.javascript.browser","version":"8.19.0","packages":[{"name":"npm:@sentry/browser","version":"8.19.0"}]},"contexts":{"trace":{"trace_id":"c0a2d394093b4f578d140afe81cbb141","span_id":"9ce7001290457fcc"},"TypeError":{}},"extra":{"appState":{"browser":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36","version":"11.16.15","persistedState":{"data":{"AuthenticationController":{"isSignedIn":"boolean"},"UserStorageController":{"isProfileSyncingEnabled":true},"MetamaskNotificationsController":{"subscriptionAccountsSeen":"object","isFeatureAnnouncementsEnabled":"boolean","isMetamaskNotificationsEnabled":"boolean","isMetamaskNotificationsFeatureSeen":"boolean","metamaskNotificationsList":"object","metamaskNotificationsReadList":"object"},"AccountsController":{"internalAccounts":{"selectedAccount":"string","accounts":"object"}},"AlertController":{"alertEnabledness":{"unconnectedAccount":true,"web3ShimUsage":true},"unconnectedAccountAlertShownOrigins":"object","web3ShimUsageOrigins":"object"},"AnnouncementController":{"announcements":"object"},"NetworkOrderController":{"orderedNetworkList":{"0":"object","1":"object","2":"object"}},"AccountOrderController":{"pinnedAccountList":{},"hiddenAccountList":{}},"AppStateController":{"browserEnvironment":{},"nftsDropdownState":{},"connectedStatusPopoverHasBeenShown":true,"termsOfUseLastAgreed":1722323809392,"defaultHomeActiveTabName":null,"fullScreenGasPollTokens":[],"notificationGasPollTokens":[],"popupGasPollTokens":[],"qrHardware":{},"recoveryPhraseReminderHasBeenShown":true,"recoveryPhraseReminderLastShown":1722323809392,"showTestnetMessageInDropdown":true,"trezorModel":null,"newPrivacyPolicyToastClickedOrClosed":"boolean","newPrivacyPolicyToastShownDate":"number","usedNetworks":{"0x1":true,"0xe708":true,"0x5":true,"0x539":true},"snapsInstallPrivacyWarningShown":true},"CurrencyController":{"currentCurrency":"usd","currencyRates":{"ETH":{"conversionDate":1665507600,"conversionRate":1700,"usdConversionRate":1700}}},"GasFeeController":{"estimatedGasFeeTimeBounds":{},"gasEstimateType":"none","gasFeeEstimates":{}},"KeyringController":{"vault":"string"},"MetaMetricsController":{"eventsBeforeMetricsOptIn":"object","fragments":"object","metaMetricsId":"fake-metrics-id","participateInMetaMetrics":true,"dataCollectionForMarketing":"boolean","traits":"object"},"NetworkController":{"selectedNetworkClientId":"string","networksMetadata":{"networkConfigurationId":{"EIPS":{},"status":"available"}},"providerConfig":{"chainId":"0x539","nickname":"Localhost 8545","rpcPrefs":"object","rpcUrl":"string","ticker":"ETH","type":"rpc","id":"networkConfigurationId"},"networkConfigurations":"object"},"OnboardingController":{"completedOnboarding":true,"firstTimeFlowType":"import","onboardingTabs":"object","seedPhraseBackedUp":true},"PermissionController":{"subjects":"object"},"PreferencesController":{"advancedGasFee":null,"currentLocale":"en","useExternalServices":"boolean","dismissSeedBackUpReminder":true,"featureFlags":{},"forgottenPassword":false,"identities":"object","ipfsGateway":"string","knownMethodData":"object","ledgerTransportType":"webhid","lostIdentities":"object","openSeaEnabled":false,"preferences":{"hideZeroBalanceTokens":false,"showExtensionInFullSizeView":false,"showFiatInTestnets":false,"showTestNetworks":false,"smartTransactionsOptInStatus":false,"useNativeCurrencyAsPrimaryCurrency":true,"petnamesEnabled":true,"showTokenAutodetectModal":"boolean","isRedesignedConfirmationsDeveloperEnabled":"boolean","showConfirmationAdvancedDetails":false},"selectedAddress":"string","theme":"light","useBlockie":false,"useNftDetection":false,"useNonceField":false,"usePhishDetect":true,"useTokenDetection":false,"useCurrencyRateCheck":true,"useMultiAccountBalanceChecker":true,"useRequestQueue":true},"QueuedRequestController":{"queuedRequestCount":0},"SelectedNetworkController":{"domains":"object"},"SmartTransactionsController":{"smartTransactionsState":{"fees":{},"liveness":true,"smartTransactions":"object"}},"SubjectMetadataController":{"subjectMetadata":"object"},"TokensController":{"allDetectedTokens":{},"allIgnoredTokens":{},"allTokens":{},"detectedTokens":"object","ignoredTokens":"object","tokens":"object"},"TransactionController":{"transactions":"object"},"config":"object","firstTimeInfo":"object"}}}}}
```

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/26216?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/26217

## **Manual testing steps**

1. Check ci

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: test/e2e/tests/metrics/errors.spec.js [hunk 1]
```diff
@@ -317,12 +317,10 @@ describe('Sentry errors', function () {
...
           const mockJsonBody = JSON.parse(mockTextBody[2]);
-          const { level } = mockJsonBody;
-          const [{ type, value }] = mockJsonBody.exception.values;
           // Verify request
-          assert.equal(type, 'TypeError');
-          assert(value.includes(migrationError));
-          assert.equal(level, 'error');
+          assert(
+            JSON.stringify(mockJsonBody.exception).includes(migrationError),
+          );
```
Comment (seaona) [inline test/e2e/tests/metrics/errors.spec.js:323]: this wouldn't be needed, as if we reach this part it means that the expected error is already captured in the mock, but we leave it for debugging purposes

**Review by DDDDDanica — DISMISSED** [review]

**Comment by DDDDDanica** [comment]
Good catch !

File: test/e2e/tests/metrics/errors.spec.js (hunk 1 above)
Comment (Gudahtt) [reply test/e2e/tests/metrics/errors.spec.js:323]: Hmm, I am not sure what this comment means, could you elaborate? This does appear to be the only assertion.

Comment (Gudahtt) [inline test/e2e/tests/metrics/errors.spec.js:323]: This assertion would probably just resolve to true or false, providing little debugging information upon failure.

Perhaps we could use` assert.match` instead:

```suggestion
          assert.match(
            JSON.stringify(mockJsonBody.exception),
            RegExp(migrationError).
          );
```

Comment (seaona) [reply test/e2e/tests/metrics/errors.spec.js:323]: yes, so a couple of lines above, we are waiting for our mock request to be completed, 

```
await driver.wait(async () => {
   const isPending = await mockedEndpoint.isPending();
   return isPending === false;
}, 3000);
const [mockedRequest] = await mockedEndpoint.getSeenRequests();
```
then, we check that the mock request body is the one we expect, but this will be already true if the condition above is true, because it means that it has the body we state above, the reason why is intercepted.

The only case the assertion won't be true, is how we had it before (where we check separately the error type and the body, and in the case that an error included the body we were looking for, but the error type was different - which is what happened in the flakiness).

So basically I think that this assert is a bit redundant, but we can also leave it. Does it make sense?

Comment (Gudahtt) [reply test/e2e/tests/metrics/errors.spec.js:323]: Ohhhh I see, sorry I missed how the test mock worked, that makes sense!

Comment (Gudahtt) [reply test/e2e/tests/metrics/errors.spec.js:323]: Given your explanation of how the test mock works, this is no longer important

**Review by Gudahtt — DISMISSED** [review]
LGTM!

**Comment by hjetpoluru** [comment]
@seaona, looks like the lint is failing.

**Comment by seaona** [comment]
thank you @hjetpoluru ! It should be fixed now :+1:

**Review by Gudahtt — DISMISSED** [review]

**Review by hjetpoluru — APPROVED** [review]
LGTM

**Review by HowardBraham — APPROVED** [review]
